File: src/main/java/cz/muni/fi/pa039/Sketch.java
package cz.muni.fi.pa039;

import cz.muni.fi.pa039.widgets.AbstractWidget;
import cz.muni.fi.pa039.widgets.ButtonWidget;
import cz.muni.fi.pa039.widgets.InputFieldWidget;
import cz.muni.fi.pa039.widgets.PointWidget;
import cz.muni.fi.pa039.widgets.TextWidget;
import processing.core.PApplet;

import java.util.HashSet;
import java.util.Optional;
import java.util.Random;
import java.util.Set;

/**
 * TODO:
 * -> CLEAR ALL button
 */
public class Sketch extends PApplet {

    private final Random random = new Random();

    private final Set<PointWidget> points = new HashSet<>();
    private final Set<AbstractWidget> widgets = new HashSet<>();

    /* SWITCHES */
    private boolean isInPointsEditingMode = true;
    private boolean giftWrappingEnabled = false;
    private boolean grahamScanEnabled = false;

    private String textToWrite = "";

    /**
     * A point we are currently dragging
     */
    private PointWidget currentDraggingPoint = null;
    /**
     * This variable serves as text input handler! When this is not null, all text input goes here
     */
    private InputFieldWidget selectedInputField = null;
    /**
     * Helper field to indicate that a button was selected
     */
    private ButtonWidget selectedButton = null;

    private int diameter = 12;
    private int windowSize = 800;
    private int leftPanelSize = 200;

    private void handleWidgetClick(AbstractWidget widget) {
        widget.setSelected(true);
        if (widget instanceof ButtonWidget) {
            selectedButton = (ButtonWidget) widget;
            handleButtonOnClickAction(widget.getId());
        }
        if (widget instanceof InputFieldWidget) {
            setSelectedInputField((InputFieldWidget) widget);
            return;
        }
        disableSelectedInputField();
    }

    private void handleButtonOnClickAction(int widgetId) {
        switch (widgetId) {
            case Ids.RANDOM_POINTS_GENERATE_BUTTON:
                generateRandomPoints();
                break;
        }
    }

    private void generateRandomPoints() {
        for (AbstractWidget w : widgets) {
            if (w.getId() == Ids.RANDOM_POINTS_INPUT_FIELD) {
                InputFieldWidget ifw = (InputFieldWidget) w;
                int amount;
                try {
                    amount = Integer.parseInt(ifw.text);
                } catch (Exception ex) {
                    amount = 0;
                }
                for (int i = 0; i < amount; i++) {
                    points.add(new PointWidget(randomPoint(), diameter / 2));
                }
            }
        }
    }


    public void settings() {
        fullScreen();
//        size(windowSize, windowSize);
    }

    public void setup() {
        background(255);
        fill(135, 175, 163);
        rect(0, 0, leftPanelSize, windowSize);
        stroke(0);
        fill(0);
        widgets.add(new TextWidget(new Point(25, 45), "Number of random points:", Ids.RANDOM_POINTS_TITLE_TEXT));
        widgets.add(new InputFieldWidget(new Point(25, 55), 50, 30, textToWrite, Ids.RANDOM_POINTS_INPUT_FIELD));
        widgets.add(new ButtonWidget(new Point(80, 55), 90, 30, "GENERATE!", Ids.RANDOM_POINTS_GENERATE_BUTTON));
    }

    public void draw() {
        background(255);
        fill(150, 241, 255);
        stroke(150, 241, 255);
        rect(0, 0, leftPanelSize, windowSize);
        fill(255, 0, 0);
        text(isInPointsEditingMode ? "Points editing ENABLED." : "Points editing DISABLED.", 25, 25);
        textToWrite = String.valueOf(System.currentTimeMillis());
        if (grahamScanEnabled && points.size() > 2) {
            performGrahamScan();
        }

        if (giftWrappingEnabled && points.size() > 2) {
            performGiftWrapping();
        }

        points.stream()
                .forEach(p -> {
                    fill(0);
                    stroke(0);
                    ellipse(p.point.x, p.point.y, p.radius * 2, p.radius * 2);
                });

        widgets.stream()
                .forEach(this::drawWidget);
    }

    private void performGiftWrapping() {

    }

    private void performGrahamScan() {

    }

    /* MOUSE EVENTS */

    public void mousePressed() {
        if (isInPointsEditingMode) {
            if (mouseButton == RIGHT) {
                deletePointContaining(mouseX, mouseY);
            } else {
                PointWidget found = findPointContaining(mouseX, mouseY);
                if (found == null) {
                    points.add(new PointWidget(new Point(mouseX, mouseY), diameter / 2));
                } else {
                    currentDraggingPoint = found;
                }
            }
        } else {
            Optional<AbstractWidget> selected = widgets.stream()
                    .filter(widget -> widget.pointIsInside(mouseX, mouseY))
                    .findFirst();
            if (selected.isPresent()) {
                handleWidgetClick(selected.get());
            } else {
                disableSelectedInputField();
            }
        }
    }

    public void mouseReleased() {
        if (!isInPointsEditingMode) {
            if (selectedButton != null) {
                selectedButton.setSelected(false);
            }
            selectedButton = null;
        } else {
            currentDraggingPoint = null;
        }
    }

    public void mouseDragged() {
        if (!isInPointsEditingMode) {
            return;
        }
        if (currentDraggingPoint != null) {
            points.remove(currentDraggingPoint);
            PointWidget p = new PointWidget(new Point(mouseX, mouseY), diameter / 2);
            currentDraggingPoint = p;
            points.add(p);
        }
    }

    /* KEYBOARD EVENTS */

    public void keyPressed() {
        switch (key) {
            case RETURN:
            case ENTER:
                isInPointsEditingMode = !isInPointsEditingMode;
                break;
            case 'p':
                if (isInPointsEditingMode) {
                    points.add(new PointWidget(randomPoint(), diameter / 2));
                }
            default:
                if (selectedInputField != null) {
                    selectedInputField.text += key;
                }
        }
    }

    private Point randomPoint() {
        int x = random.nextInt(windowSize - 200 - diameter) + 200 + diameter / 2;
        int y = random.nextInt(windowSize - diameter) + diameter / 2;
        return new Point(x, y);
    }

    private void drawWidget(AbstractWidget widget) {
        if (widget instanceof InputFieldWidget) {
            drawInputField((InputFieldWidget) widget);
        } else if (widget instanceof TextWidget) {
            drawTextWidget((TextWidget) widget);
        }
    }

    private void disableSelectedInputField() {
        if (selectedInputField != null) {
            selectedInputField.setSelected(false);
        }
        selectedInputField = null;
    }

    private void setSelectedInputField(InputFieldWidget widget) {
        widget.text = "";
        selectedInputField = widget;
        selectedInputField.setSelected(true);
    }

    private PointWidget findPointContaining(int x, int y) {
        for (PointWidget p : points) {
            if (p.pointIsInside(x, y)) {
                return p;
            }
        }
        return null;
    }

    private void deletePointContaining(int x, int y) {
        points.removeIf(p -> p.pointIsInside(x, y));
    }

    private void drawInputField(InputFieldWidget inputField) {
        fill(inputField.getFill());
        stroke(inputField.getStroke());
        int topLeftX = inputField.topLeft.x;
        int topLeftY = inputField.topLeft.y;
        rect(topLeftX, topLeftY, inputField.width, inputField.height, 2);
        fill(inputField.getStroke());
        text(inputField.text, topLeftX + 5, topLeftY + inputField.height - 10);
    }

    private void drawTextWidget(TextWidget textWidget) {
        fill(0);
        stroke(0);
        text(textWidget.text, textWidget.bottomLeft.x, textWidget.bottomLeft.y);
    }
}

File: src/main/java/cz/muni/fi/pa039/Sketch.kt
package cz.muni.fi.pa039

import cz.muni.fi.pa039.widgets.AbstractWidget
import cz.muni.fi.pa039.widgets.ButtonWidget
import cz.muni.fi.pa039.widgets.InputFieldWidget
import cz.muni.fi.pa039.widgets.PointWidget
import cz.muni.fi.pa039.widgets.TextWidget
import processing.core.PApplet
import processing.core.PConstants

import java.util.HashSet
import java.util.Random

/**
 * TODO:
 * -> CLEAR ALL button
 */
class Sketch : PApplet() {

    private val random = Random()

    private val points = HashSet<PointWidget>()
    private val widgets = HashSet<AbstractWidget>()

    /* SWITCHES */
    private var isInPointsEditingMode = true
    private val giftWrappingEnabled = false
    private val grahamScanEnabled = false

    private var textToWrite = ""

    /**
     * A point we are currently dragging
     */
    private var currentDraggingPoint: PointWidget? = null
    /**
     * This variable serves as text input handler! When this is not null, all text input goes here
     */
    private var selectedInputField: InputFieldWidget? = null
    /**
     * Helper field to indicate that a button was selected
     */
    private var selectedButton: ButtonWidget? = null

    private val diameter = 12
    private val windowSize = 800
    private val leftPanelSize = 200

    private fun handleWidgetClick(widget: AbstractWidget) {
        widget.isSelected = true
        if (widget is ButtonWidget) {
            selectedButton = widget
            handleButtonOnClickAction(widget.id)
        }
        if (widget is InputFieldWidget) {
            setSelectedInputField(widget)
            return
        }
        disableSelectedInputField()
    }

    private fun handleButtonOnClickAction(widgetId: Int) {
        when (widgetId) {
            Ids.RANDOM_POINTS_GENERATE_BUTTON -> generateRandomPoints()
        }
    }

    private fun generateRandomPoints() {
        for (w in widgets) {
            if (w.id == Ids.RANDOM_POINTS_INPUT_FIELD) {
                val ifw = w as InputFieldWidget
                var amount: Int
                try {
                    amount = Integer.parseInt(ifw.text)
                } catch (ex: Exception) {
                    amount = 0
                }

                for (i in 0 until amount) {
                    points.add(PointWidget(randomPoint(), diameter / 2))
                }
            }
        }
    }


    override fun settings() {
        fullScreen()
        //        size(windowSize, windowSize);
    }

    override fun setup() {
        background(255)
        fill(135f, 175f, 163f)
        rect(0f, 0f, leftPanelSize.toFloat(), windowSize.toFloat())
        stroke(0)
        fill(0)
        widgets.add(TextWidget(Point(25, 45), "Number of random points:", Ids.RANDOM_POINTS_TITLE_TEXT))
        widgets.add(InputFieldWidget(Point(25, 55), 50, 30, textToWrite, Ids.RANDOM_POINTS_INPUT_FIELD))
        widgets.add(ButtonWidget(Point(80, 55), 90, 30, "GENERATE!", Ids.RANDOM_POINTS_GENERATE_BUTTON))
    }

    override fun draw() {
        background(255)
        fill(150f, 241f, 255f)
        stroke(150f, 241f, 255f)
        rect(0f, 0f, leftPanelSize.toFloat(), windowSize.toFloat())
        fill(255f, 0f, 0f)
        text(if (isInPointsEditingMode) "Points editing ENABLED." else "Points editing DISABLED.", 25f, 25f)
        textToWrite = System.currentTimeMillis().toString()
        if (grahamScanEnabled && points.size > 2) {
            performGrahamScan()
        }

        if (giftWrappingEnabled && points.size > 2) {
            performGiftWrapping()
        }

        points.forEach { p ->
                    fill(0)
                    stroke(0)
                    ellipse(p.point.x.toFloat(), p.point.y.toFloat(), (p.radius * 2).toFloat(), (p.radius * 2).toFloat())
                }

        widgets.forEach(this::drawWidget)
    }

    private fun performGiftWrapping() {

    }

    private fun performGrahamScan() {

    }

    /* MOUSE EVENTS */

    override fun mousePressed() {
        if (isInPointsEditingMode) {
            if (mouseButton == PConstants.RIGHT) {
                deletePointContaining(mouseX, mouseY)
            } else {
                val found = findPointContaining(mouseX, mouseY)
                if (found == null) {
                    points.add(PointWidget(Point(mouseX, mouseY), diameter / 2))
                } else {
                    currentDraggingPoint = found
                }
            }
        } else {
            val selected = widgets.stream()
                    .filter { widget -> widget.pointIsInside(mouseX, mouseY) }
                    .findFirst()
            if (selected.isPresent) {
                handleWidgetClick(selected.get())
            } else {
                disableSelectedInputField()
            }
        }
    }

    override fun mouseReleased() {
        if (!isInPointsEditingMode) {
            if (selectedButton != null) {
                selectedButton!!.isSelected = false
            }
            selectedButton = null
        } else {
            currentDraggingPoint = null
        }
    }

    override fun mouseDragged() {
        if (!isInPointsEditingMode) {
            return
        }
        if (currentDraggingPoint != null) {
            points.remove(currentDraggingPoint!!)
            val p = PointWidget(Point(mouseX, mouseY), diameter / 2)
            currentDraggingPoint = p
            points.add(p)
        }
    }

    /* KEYBOARD EVENTS */

    override fun keyPressed() {
        when (key) {
            PConstants.RETURN, PConstants.ENTER -> isInPointsEditingMode = !isInPointsEditingMode
            'p' -> {
                if (isInPointsEditingMode) {
                    points.add(PointWidget(randomPoint(), diameter / 2))
                }
                if (selectedInputField != null) {
                    selectedInputField!!.text += key
                }
            }
            else -> if (selectedInputField != null) {
                selectedInputField!!.text += key
            }
        }
    }

    private fun randomPoint(): Point {
        val x = random.nextInt(windowSize - 200 - diameter) + 200 + diameter / 2
        val y = random.nextInt(windowSize - diameter) + diameter / 2
        return Point(x, y)
    }

    private fun drawWidget(widget: AbstractWidget) {
        if (widget is InputFieldWidget) {
            drawInputField(widget)
        } else if (widget is TextWidget) {
            drawTextWidget(widget)
        }
    }

    private fun disableSelectedInputField() {
        if (selectedInputField != null) {
            selectedInputField!!.isSelected = false
        }
        selectedInputField = null
    }

    private fun setSelectedInputField(widget: InputFieldWidget) {
        widget.text = ""
        selectedInputField = widget
        selectedInputField!!.isSelected = true
    }

    private fun findPointContaining(x: Int, y: Int): PointWidget? {
        for (p in points) {
            if (p.pointIsInside(x, y)) {
                return p
            }
        }
        return null
    }

    private fun deletePointContaining(x: Int, y: Int) {
        points.removeIf { p -> p.pointIsInside(x, y) }
    }

    private fun drawInputField(inputField: InputFieldWidget) {
        fill(inputField.fill)
        stroke(inputField.stroke)
        val topLeftX = inputField.topLeft.x
        val topLeftY = inputField.topLeft.y
        rect(topLeftX.toFloat(), topLeftY.toFloat(), inputField.width.toFloat(), inputField.height.toFloat(), 2f)
        fill(inputField.stroke)
        text(inputField.text, (topLeftX + 5).toFloat(), (topLeftY + inputField.height - 10).toFloat())
    }

    private fun drawTextWidget(textWidget: TextWidget) {
        fill(0)
        stroke(0)
        text(textWidget.text, textWidget.bottomLeft.x.toFloat(), textWidget.bottomLeft.y.toFloat())
    }
}
